File: apps/apple/core/Sources/LorvexWorkflow/HabitReminderOps.swift
import Foundation
import GRDB
import LorvexDomain
import LorvexStore

/// Habit reminder policy CRUD — the canonical implementations every write
/// surface (MCP, app, CLI) delegates to instead of maintaining independent SQL.
public enum HabitReminderOps {
  /// A fully loaded habit reminder policy row, joined with the habit name.
  ///
  /// `version` is carried so callers that route through the canonical
  /// `habit_reminder_policy` payload builder emit the 7-field wire shape.
  public struct HabitReminderPolicyRow: Sendable, Equatable {
    public let id: String
    public let habitId: String
    public let habitName: String
    public let reminderTime: String
    public let enabled: Bool
    public let createdAt: String
    public let updatedAt: String
    public let version: String
  }

  private static let policySelect =
    "SELECT p.id, p.habit_id, h.name, p.reminder_time, p.enabled, p.created_at, p.updated_at, "
    + "p.version "
    + "FROM habit_reminder_policies p JOIN habits h ON h.id = p.habit_id"

  private static func rowFromQuery(_ row: Row) -> HabitReminderPolicyRow {
    HabitReminderPolicyRow(
      id: row[0], habitId: row[1], habitName: row[2], reminderTime: row[3],
      enabled: (row[4] as Int64) != 0, createdAt: row[5], updatedAt: row[6], version: row[7])
  }

  /// Load a single habit reminder policy by ID. Returns `nil` when the row
  /// does not exist; reserves a thrown error for genuine query failures.
  public static func loadPolicyById(_ db: Database, policyId: String) throws
    -> HabitReminderPolicyRow?
  {
    guard
      let row = try Row.fetchOne(
        db, sql: "\(policySelect) WHERE p.id = ?", arguments: [policyId])
    else { return nil }
    return rowFromQuery(row)
  }

  /// Load all habit reminder policies, ordered by habit name (case-insensitive)
  /// then reminder time.
  public static func listAllPolicies(_ db: Database) throws -> [HabitReminderPolicyRow] {
    let rows = try Row.fetchAll(
      db, sql: "\(policySelect) ORDER BY h.name COLLATE NOCASE ASC, p.reminder_time ASC")
    return rows.map(rowFromQuery)
  }

  /// Load the reminder policies for one habit, ordered by reminder time. Filters
  /// in SQL rather than loading every habit's policies and discarding the rest.
  public static func listPolicies(_ db: Database, habitId: String) throws
    -> [HabitReminderPolicyRow]
  {
    let rows = try Row.fetchAll(
      db, sql: "\(policySelect) WHERE p.habit_id = ? ORDER BY p.reminder_time ASC",
      arguments: [habitId])
    return rows.map(rowFromQuery)
  }

  /// ID of an existing policy that would conflict (same habit + same time),
  /// optionally excluding a specific policy ID (for updates).
  private static func loadConflictingSlotId(
    _ db: Database, habitId: String, reminderTime: String, excludingId: String?
  ) throws -> String? {
    if let excludingId = excludingId {
      return try String.fetchOne(
        db,
        sql: "SELECT id FROM habit_reminder_policies "
          + "WHERE habit_id = ? AND reminder_time = ? AND id != ?",
        arguments: [habitId, reminderTime, excludingId])
    }
    return try String.fetchOne(
      db,
      sql: "SELECT id FROM habit_reminder_policies WHERE habit_id = ? AND reminder_time = ?",
      arguments: [habitId, reminderTime])
  }

  /// Parameters for upserting a habit reminder policy. `policyId` `nil` or
  /// blank means create new; the caller supplies `version` and `now`.
  public struct UpsertParams: Sendable {
    public let policyId: String?
    public let habitId: String
    public let reminderTime: String
    public let enabled: Bool
    public let version: String
    public let now: String

    public init(
      policyId: String?, habitId: String, reminderTime: String, enabled: Bool,
      version: String, now: String
    ) {
      self.policyId = policyId
      self.habitId = habitId
      self.reminderTime = reminderTime
      self.enabled = enabled
      self.version = version
      self.now = now
    }
  }

  /// Create or update a habit reminder policy.
  ///
  /// Validates that `habit_id` is non-empty and references an existing habit,
  /// `reminder_time` is valid `HH:MM`, no other slot exists at the same time
  /// for the same habit, and (on update) the slot belongs to the named habit.
  /// Returns the fully loaded policy row.
  public static func upsertHabitReminderPolicy(_ db: Database, params: UpsertParams) throws
    -> HabitReminderPolicyRow
  {
    let habitId = params.habitId.trimmingCharacters(in: .whitespaces)
    if habitId.isEmpty {
      throw StoreError.validation("habit_id must not be empty")
    }

    if case .failure(let e) = ValidationFormat.validateTimeFormat(params.reminderTime) {
      throw StoreError.validation(
        "invalid reminder_time '\(params.reminderTime)': \(e.description)")
    }

    let habitExists = try String.fetchOne(
      db, sql: "SELECT id FROM habits WHERE id = ?", arguments: [habitId])
    if habitExists == nil {
      throw StoreError.notFound(entity: "habit", id: habitId)
    }

    let enabledVal: Int64 = params.enabled ? 1 : 0

    let resolvedPolicyId: String
    let trimmedPolicyId = params.policyId?.trimmingCharacters(in: .whitespaces)
    if let id = trimmedPolicyId, !id.isEmpty {
      // UPDATE path.
      guard
        let existing = try Row.fetchOne(
          db,
          sql: "SELECT habit_id FROM habit_reminder_policies WHERE id = ?",
          arguments: [id])
      else {
        throw StoreError.notFound(entity: "habit_reminder_policy", id: id)
      }
      let existingHabitId: String = existing["habit_id"]
      if existingHabitId != habitId {
        throw StoreError.validation(
          "habit reminder slot '\(id)' belongs to a different habit")
      }
      if try loadConflictingSlotId(
        db, habitId: habitId, reminderTime: params.reminderTime, excludingId: id) != nil
      {
        throw StoreError.validation(
          "habit '\(habitId)' already has a reminder slot at \(params.reminderTime)")
      }
      try db.execute(
        sql: "UPDATE habit_reminder_policies "
          + "SET reminder_time = ?, enabled = ?, version = ?, updated_at = ? "
          + "WHERE id = ? AND version < ?",
        arguments: [
          params.reminderTime, enabledVal, params.version, params.now, id, params.version,
        ])
      if db.changesCount == 0 {
        let observed = try String.fetchOne(
          db, sql: "SELECT version FROM habit_reminder_policies WHERE id = ?", arguments: [id])
        guard let observed else {
          throw StoreError.notFound(entity: EntityName.habitReminderPolicy, id: id)
        }
        throw StoreError.versionSuperseded(
          entityType: EntityName.habitReminderPolicy, entityId: id,
          attemptedVersion: params.version,
          existingVersion: observed)
      }
      resolvedPolicyId = id
    } else {
      // INSERT path.
      if try loadConflictingSlotId(
        db, habitId: habitId, reminderTime: params.reminderTime, excludingId: nil) != nil
      {
        throw StoreError.validation(
          "habit '\(habitId)' already has a reminder slot at \(params.reminderTime)")
      }
      let id = EntityID.newEntityIDString()
      try db.execute(
        sql: "INSERT INTO habit_reminder_policies "
          + "(id, habit_id, reminder_time, enabled, version, created_at, updated_at) "
          + "VALUES (?, ?, ?, ?, ?, ?, ?)",
        arguments: [
          id, habitId, params.reminderTime, enabledVal, params.version, params.now, params.now,
        ])
      resolvedPolicyId = id
    }

    guard let row = try loadPolicyById(db, policyId: resolvedPolicyId) else {
      throw StoreError.validation(
        "habit reminder policy '\(resolvedPolicyId)' not found after upsert")
    }
    return row
  }

  /// Result of deleting a habit reminder policy.
  public struct DeleteResult: Sendable, Equatable {
    public let deleted: Bool
  }

  /// Delete a habit reminder policy by ID. Returns whether a row was removed.
  public static func deleteHabitReminderPolicy(_ db: Database, policyId: String) throws
    -> DeleteResult
  {
    try db.execute(
      sql: "DELETE FROM habit_reminder_policies WHERE id = ?", arguments: [policyId])
    return DeleteResult(deleted: db.changesCount > 0)
  }

  /// Stamp `habit_reminder_delivery_state.last_delivered_at` for `policyId` so the
  /// occurrence planner debounces same-period re-nudges. Device-local (never
  /// synced). `deliveredAt` is the instant the most recent in-period occurrence
  /// elapsed; `now` is the write timestamp. The deterministic device-local
  /// analog of an OS delivery callback (Apple pre-schedules notifications the
  /// system fires without a reliable callback). Callers gate this on the
  /// policy's `last_armed_at`: only an occurrence this device actually armed
  /// with the OS can be recorded as delivered.
  public static func markHabitReminderDelivered(
    _ db: Database, policyId: String, deliveredAt: String, now: String
  ) throws {
    try db.execute(
      sql: """
        INSERT INTO habit_reminder_delivery_state \
            (policy_id, last_delivered_at, updated_at) \
        VALUES (?1, ?2, ?3) \
        ON CONFLICT(policy_id) DO UPDATE SET \
          last_delivered_at = excluded.last_delivered_at, \
          updated_at = excluded.updated_at
        """,
      arguments: [policyId, deliveredAt, now])
  }

  /// Replace the device's armed-occurrence record with exactly the occurrences
  /// the habit notification scheduler reported as accepted this pass.
  ///
  /// `armedThroughByPolicy` maps a policy id to the latest occurrence fire
  /// time (sync-timestamp string) the OS accepted a request for; the armed set
  /// per policy is the contiguous earliest prefix of its occurrences, so one
  /// "armed through" instant fully describes it. Two writes keep
  /// `last_armed_at` a mirror of the currently pending
  /// `UNUserNotificationCenter` request set, which every replace pass rebuilds
  /// from scratch:
  /// - each mapped policy is stamped with its armed-through instant;
  /// - every other policy's `last_armed_at` is cleared back to NULL, because
  ///   the replace pass just removed its OS requests (budgeted out, permission
  ///   denied, or add failed).
  ///
  /// The delivered reconciler only records occurrences at or before
  /// `last_armed_at`, so clearing keeps a dropped nudge visible as due instead
  /// of letting a stale armed stamp record a phantom delivery.
  public static func replaceHabitRemindersArmed(
    _ db: Database, armedThroughByPolicy: [String: String], now: String
  ) throws {
    let armedPolicyIds = Array(armedThroughByPolicy.keys).sorted()
    let placeholders = Array(
      repeating: "?", count: armedPolicyIds.count
    ).joined(separator: ", ")
    let notInClause =
      armedPolicyIds.isEmpty ? "" : "AND policy_id NOT IN (\(placeholders)) "
    try db.execute(
      sql: """
        UPDATE habit_reminder_delivery_state SET \
          last_armed_at = NULL, \
          updated_at = ?1 \
        WHERE last_armed_at IS NOT NULL \
          \(notInClause)
        """,
      arguments: StatementArguments([now] + armedPolicyIds))
    for policyId in armedPolicyIds {
      guard let armedThrough = armedThroughByPolicy[policyId] else { continue }
      try db.execute(
        sql: """
          INSERT INTO habit_reminder_delivery_state \
              (policy_id, last_armed_at, updated_at) \
          VALUES (?1, ?2, ?3) \
          ON CONFLICT(policy_id) DO UPDATE SET \
            last_armed_at = excluded.last_armed_at, \
            updated_at = excluded.updated_at
          """,
        arguments: [policyId, armedThrough, now])
    }
  }
}

File: apps/tauri/lorvex-workflow/src/habit_reminder_ops.rs
//! Shared habit reminder policy mutation operations.
//!
//! These are the canonical implementations for habit reminder policy CRUD.
//! Both MCP and Tauri delegate to these instead of maintaining independent SQL.

use lorvex_domain::validation::validate_time_format;
use rusqlite::{params, Connection, OptionalExtension};
use serde::Serialize;

use lorvex_store::StoreError;

// ---------------------------------------------------------------------------
// Result types
// ---------------------------------------------------------------------------

/// A fully loaded habit reminder policy row (joined with the habit name).
///
/// `version` is included so callers that route through the
/// `lorvex_store::payload_loaders::habit_reminder_policy_payload`
/// primitive emit the canonical 7-field wire shape;
/// hand-rolled audit payload dropped `version`, drifting from every
/// other surface.
#[derive(Debug, Clone, Serialize)]
pub struct HabitReminderPolicyRow {
    pub id: String,
    pub habit_id: String,
    pub habit_name: String,
    pub reminder_time: String,
    pub enabled: bool,
    pub created_at: String,
    pub updated_at: String,
    pub version: String,
}

fn row_from_query(row: &rusqlite::Row<'_>) -> rusqlite::Result<HabitReminderPolicyRow> {
    Ok(HabitReminderPolicyRow {
        id: row.get(0)?,
        habit_id: row.get(1)?,
        habit_name: row.get(2)?,
        reminder_time: row.get(3)?,
        enabled: row.get(4)?,
        created_at: row.get(5)?,
        updated_at: row.get(6)?,
        version: row.get(7)?,
    })
}

const POLICY_SELECT: &str = "\
    SELECT p.id, p.habit_id, h.name, p.reminder_time, p.enabled, p.created_at, p.updated_at, \
           p.version \
    FROM habit_reminder_policies p JOIN habits h ON h.id = p.habit_id";

// ---------------------------------------------------------------------------
// Read helpers
// ---------------------------------------------------------------------------

/// Load a single habit reminder policy by ID. Returns `Ok(None)` if the
/// row does not exist; reserve `StoreError` for genuine query failures
/// so callers can distinguish "missing" from "broken".
pub fn load_policy_by_id(
    conn: &Connection,
    policy_id: &str,
) -> Result<Option<HabitReminderPolicyRow>, StoreError> {
    use rusqlite::OptionalExtension as _;
    static SQL: std::sync::OnceLock<String> = std::sync::OnceLock::new();
    conn.prepare_cached(SQL.get_or_init(|| format!("{POLICY_SELECT} WHERE p.id = ?1")))?
        .query_row(params![policy_id], row_from_query)
        .optional()
        .map_err(StoreError::from)
}

/// Load all habit reminder policies, ordered by habit name then time.
pub fn list_all_policies(conn: &Connection) -> Result<Vec<HabitReminderPolicyRow>, StoreError> {
    static SQL: std::sync::OnceLock<String> = std::sync::OnceLock::new();
    let sql = SQL.get_or_init(|| {
        format!("{POLICY_SELECT} ORDER BY h.name COLLATE NOCASE ASC, p.reminder_time ASC")
    });
    let mut stmt = conn.prepare(sql)?;
    let policies = stmt
        .query_map([], row_from_query)?
        .collect::<Result<Vec<_>, _>>()?;
    Ok(policies)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/// Returns the ID of an existing policy that would conflict (same habit +
/// same time), optionally excluding a specific policy ID (for updates).
fn load_conflicting_slot_id(
    conn: &Connection,
    habit_id: &lorvex_domain::HabitId,
    reminder_time: &str,
    excluding_id: Option<&str>,
) -> Result<Option<String>, StoreError> {
    excluding_id.map_or_else(
        || {
            conn.query_row(
                "SELECT id FROM habit_reminder_policies \
                 WHERE habit_id = ?1 AND reminder_time = ?2",
                params![habit_id, reminder_time],
                |row| row.get(0),
            )
            .optional()
            .map_err(StoreError::from)
        },
        |policy_id| {
            conn.query_row(
                "SELECT id FROM habit_reminder_policies \
                 WHERE habit_id = ?1 AND reminder_time = ?2 AND id != ?3",
                params![habit_id, reminder_time, policy_id],
                |row| row.get(0),
            )
            .optional()
            .map_err(StoreError::from)
        },
    )
}

// ---------------------------------------------------------------------------
// Upsert
// ---------------------------------------------------------------------------

/// Parameters for upserting a habit reminder policy.
#[derive(Debug)]
pub struct UpsertHabitReminderPolicyParams<'a> {
    /// Optional existing policy ID. `None` or blank string means create new.
    pub policy_id: Option<&'a str>,
    pub habit_id: &'a str,
    pub reminder_time: &'a str,
    pub enabled: bool,
    pub version: &'a str,
    pub now: &'a str,
}

/// Create or update a habit reminder policy.
///
/// Validates:
/// - `habit_id` is non-empty and references an existing habit
/// - `reminder_time` is valid HH:MM (via domain `validate_time_format`)
/// - No conflicting slot at the same time for the same habit
/// - If updating, the existing slot belongs to the specified habit
///
/// Returns the fully loaded policy row on success.
pub fn upsert_habit_reminder_policy(
    conn: &Connection,
    params: &UpsertHabitReminderPolicyParams<'_>,
) -> Result<HabitReminderPolicyRow, StoreError> {
    let habit_id = params.habit_id.trim();
    if habit_id.is_empty() {
        return Err(StoreError::Validation(
            "habit_id must not be empty".to_string(),
        ));
    }

    // Validate time format using the canonical domain validator.
    validate_time_format(params.reminder_time).map_err(|e| {
        StoreError::Validation(format!(
            "invalid reminder_time '{}': {}",
            params.reminder_time, e
        ))
    })?;

    // Verify the habit exists.
    let habit_exists: Option<String> = conn
        .query_row(
            "SELECT id FROM habits WHERE id = ?1",
            params![habit_id],
            |row| row.get(0),
        )
        .optional()?;
    if habit_exists.is_none() {
        return Err(StoreError::NotFound {
            entity: "habit",
            id: habit_id.to_string(),
        });
    }

    let enabled_val: i64 = i64::from(params.enabled);

    let resolved_policy_id = if let Some(id) =
        params.policy_id.map(str::trim).filter(|v| !v.is_empty())
    {
        // --- UPDATE path ---
        let existing_habit_id: Option<String> = conn
            .query_row(
                "SELECT habit_id FROM habit_reminder_policies WHERE id = ?1",
                rusqlite::params![id],
                |row| row.get(0),
            )
            .optional()?;
        let Some(existing_habit_id) = existing_habit_id else {
            return Err(StoreError::NotFound {
                entity: "habit_reminder_policy",
                id: id.to_string(),
            });
        };
        if existing_habit_id != habit_id {
            return Err(StoreError::Validation(format!(
                "habit reminder slot '{id}' belongs to a different habit"
            )));
        }
        let typed_habit_id = lorvex_domain::HabitId::from_trusted(habit_id.to_string());
        if load_conflicting_slot_id(conn, &typed_habit_id, params.reminder_time, Some(id))?
            .is_some()
        {
            return Err(StoreError::Validation(format!(
                "habit '{habit_id}' already has a reminder slot at {}",
                params.reminder_time
            )));
        }
        conn.prepare_cached(
            "UPDATE habit_reminder_policies \
             SET reminder_time = ?1, enabled = ?2, version = ?3, updated_at = ?4 \
             WHERE id = ?5",
        )?
        .execute(rusqlite::params![
            params.reminder_time,
            enabled_val,
            params.version,
            params.now,
            id
        ])?;
        id.to_string()
    } else {
        // --- INSERT path ---
        let typed_habit_id = lorvex_domain::HabitId::from_trusted(habit_id.to_string());
        if load_conflicting_slot_id(conn, &typed_habit_id, params.reminder_time, None)?.is_some() {
            return Err(StoreError::Validation(format!(
                "habit '{habit_id}' already has a reminder slot at {}",
                params.reminder_time
            )));
        }
        let id = lorvex_domain::entity_id::new_entity_id_string();
        conn.prepare_cached(
            "INSERT INTO habit_reminder_policies \
             (id, habit_id, reminder_time, enabled, version, created_at, updated_at) \
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?6)",
        )?
        .execute(rusqlite::params![
            id,
            habit_id,
            params.reminder_time,
            enabled_val,
            params.version,
            params.now
        ])?;
        id
    };

    // After upsert the row must exist; surface a typed error if the
    // row vanished between our INSERT/UPDATE and this re-read.
    load_policy_by_id(conn, &resolved_policy_id)?.ok_or_else(|| {
        StoreError::Validation(format!(
            "habit reminder policy '{resolved_policy_id}' not found after upsert"
        ))
    })
}

// ---------------------------------------------------------------------------
// Delete
// ---------------------------------------------------------------------------

/// Result of deleting a habit reminder policy.
#[derive(Debug)]
pub struct DeleteHabitReminderPolicyResult {
    /// Whether a row was actually deleted.
    pub deleted: bool,
}

/// Delete a habit reminder policy by ID.
///
/// Returns whether a row was actually deleted.
pub fn delete_habit_reminder_policy(
    conn: &Connection,
    policy_id: &str,
) -> Result<DeleteHabitReminderPolicyResult, StoreError> {
    let deleted = conn
        .prepare_cached("DELETE FROM habit_reminder_policies WHERE id = ?1")?
        .execute(params![policy_id])?;
    Ok(DeleteHabitReminderPolicyResult {
        deleted: deleted > 0,
    })
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests;
